icicle-compiler/test/Icicle/Test/Foreign/Array.hs

```diff
@@ -251,37 +245,33 @@ univalue ty = case ty of
   IUnit  -> True
   _      -> False
 
-runWith :: Input -> (WordPtr -> IO Property) -> IO Property
-runWith (Input _ val) prop = go val
-  where
-    go xs = bracket (pokeArr xs) snd (prop . fst)
-
-runProp :: Input -> Text -> (Ptr a -> [Argument]) -> IO Property
+runProp :: Input -> Text -> (Mempool.Mempool -> Argument -> IO [Argument]) -> IO Property
 runProp input fun args = do
-  -- let ty = inputType input
   let fs = seprops input
-  runWith input $ \wptr -> runRight $ do
-    let ptr = wordPtrToPtr wptr
-    src <- readLibraryWith ["-DICICLE_ASSERT=1"] fs
-    f   <- function src fun retBool
-    r   <- liftIO $ f $ args ptr
-    -- arr <- liftIO $ peekArr ty wptr
-    -- return $ counterexample ("array after test:\n" <> show arr)
+  bracket (liftIO Mempool.create) (liftIO . Mempool.free) $ \pool -> runRight $ do
+  -- Cache the output, because we're likely to reuse the same type in another test 
+  bracketEitherT' (compileLibrary CacheLibrary ["-DICICLE_ASSERT=1"] fs) releaseLibrary $ \lib -> runRight $ do
```

Comment:
Is it OK to use CacheLibrary here? There are only seven base types, so most of the time there should be relatively few instances of the C code, but nested arrays mean there could be lots in the worst case...